### PR TITLE
First stab at using Jed's petsc docker image.

### PR DESCRIPTION
**Pull request by jeff-cohere**
Possibly another "noisy" PR. This is an attempt to speed up our builds by using a pre-built PETSc library with all the trimmings.

### CODE REVIEW

**Comment by jeff-cohere** [comment]
Hey @jedbrown . It looks like the action fails trying to write to the `.git` directory within the GitHub workspace. I think this might have to do with the fact that you define a `petsc` user with Docker's `USER` command in your image. I read somewhere in the GitHub Actions docs that you can't use the `USER` command in a Dockerfile used with an Action--specifically, that a non-root user can't write to the workspace.

Have you used this image successfully in a GitHub Action like this? If not, maybe we can create an image that's just like this one but excludes the `petsc` user part.

**Comment by jedbrown** [comment]
I wasn't aware of that constraint on GH Actions. We use `jedbrown/petsc-root` on Azure for that reason. (We were on Azure DevOps long before GH Actions and never moved it.)

File: .github/workflows/auto_pr_test.yml [hunk 1]
```diff
@@ -15,67 +15,33 @@ jobs:
   build:
     runs-on: ${{ matrix.os }}
 
+    # We use a pre-fab container with PETSc and MPICH included.
+    # NOTE: Docker containers may only work for Linux builds currently.
+    container:
+      image: jedbrown/petsc:latest
```
Comment (jedbrown) [inline .github/workflows/auto_pr_test.yml:21]:
```suggestion
      image: jedbrown/petsc-root:latest
```

File: .github/workflows/auto_pr_test.yml [hunk 2]
```diff
@@ -15,67 +15,33 @@ jobs:
...
+      env:
+        # These PETSC variables depend where PETSc is installed
+        # See https://github.com/jedbrown/jedbrown-dockerfiles/blob/main/petsc/Dockerfile
+        PETSC_DIR: /usr/local/petsc
+        PETSC_ARCH: mpich-int32-real-opt
```
Comment (jedbrown) [inline .github/workflows/auto_pr_test.yml:26]: These environment variables are exported in the container so I'd assume this job would inherit them. Also, it's a prefix install so `PETSC_ARCH` should be empty and `PETSC_DIR=/usr/loca/petsc/mpich-int32-real-opt`.

Comment (jeff-cohere) [reply .github/workflows/auto_pr_test.yml:26]: Cool, I'll remove them. And thanks for the petsc-root image.

**Comment by jeff-cohere** [comment]
Yikes. Looks like we're asking a bit much of GFortran. Looks like the failure could have something to do with the code coverage flags:

```
/usr/local/bin/mpif90 -o transient_mpfaof90.o -c -fPIC -Wall -ffree-line-length-0 -Wno-unused-dummy-argument -O2 -march=native -ffp-contract=fast   -lgcov -I/usr/local/petsc/mpich-int32-real-opt/include -I/usr/local/include   -fprofile-arcs -ftest-coverage -I/__w/TDycore/TDycore/arch-default/include -I/__w/TDycore/TDycore/include `pwd`/transient_mpfaof90.F90
during IPA pass: profile
/__w/TDycore/TDycore/demo/transient/transient_mpfaof90.F90:50:0:
   50 |   use tdycore
internal compiler error: in coverage_begin_function, at coverage.c:656
Please submit a full bug report
```

Do you happen to know what version of gfortran is used in your container? I suppose we can see whether it works without Fortran code coverage, but that's a bummer.

**Comment by jedbrown** [comment]
LOL, internal compiler error. I could spin up an image with ubuntu:20.10 (gcc-10) if that would help. I haven't encountered this before, but I suppose I haven't been using the container for coverage of Fortran modules.

**Comment by jeff-cohere** [comment]
Yeah, this is unfortunately congruent with my experiences in "Fortran compiler quality." If it's not much trouble to fab up an image with gcc-10, it'd be great to give it a try. Appreciated!

**Comment by jedbrown** [comment]
```
$ docker run --rm jedbrown/petsc-root gfortran --version
GNU Fortran (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

I'm setting up a `jedbrown/petsc-root:rolling`.

**Comment by jedbrown** [comment]
Well, damn. The bug persists in gfortran-10. What are you using locally? Or we can de-tune the coverage options. This should be reported upstream if it hasn't yet been.

**Comment by jeff-cohere** [comment]
Nuts.

I'm using gcc 10.2 locally, so I assume it would occur there, too. For now, I wouldn't mind skipping code coverage options (maybe even on a per-test basis if it's really sensitive) for now. I'll poke around with that, and also see what I can do about putting together a bug report.

Thanks for the effort!

**Comment by jeff-cohere** [comment]
Looks like a similar bug report has already been filed, with very little interest: https://bugzilla.redhat.com/show_bug.cgi?id=1722577

I'm going to see if I can get this to build by selectively removing the code coverage flags from offending Fortran source files.

**Comment by jedbrown** [comment]
Can you reproduce by running coverage locally (without Docker)?

The second bug report you linked says it's fixed upstream -- we could put a fresh build in the container. The slow action is the natural consequence of lack of paying Fortran users (who hire or contract compiler devs).

**Comment by jeff-cohere** [comment]
I can reproduce locally. Interestingly enough, I can get the F90 demos to build by excluding the profiling flags, but then 2 of 4 of the F90 MPFAO regression tests fail, and I can't reproduce that. Something fishy is going on.

Yeah, Fortran compilers are really a niche market in this day and age. Do you want to stick a new GFortran build in the container and we can try again? I can back out the changes I made to exclude the build flags from the affected demo exes. Thanks!

**Comment by jedbrown** [comment]
Is there a stand-alone test available? I wonder if it's fixed in the 10.2.1 (pre-release?) on Debian sid: https://packages.debian.org/sid/gfortran-10

**Comment by jeff-cohere** [comment]
Yes, you can reproduce from a clean build with the following from the top of your source tree:

```
make -j codecov=1 V=1
cd demo
make -j codecov=1 V=1
cd ../regression_tests
make test-transient-mpfaof90
```

Let me know if that doesn't give you an internal compiler error on GFortran 10.2.

**Comment by jeff-cohere** [comment]
The fix might be in that new Debian release. It was cherry-picked on Nov 12 to gcc-10, and there seem to be items in the change list that come after that date.

**Comment by jedbrown** [comment]
Fussy baby or I'd be hands on keyboard. If it's an ICE, then it's at compile-time, not run-time. Does it trigger if we just preprocess the file and then try to compile that with coverage? That would save needing to install PETSc to test a compiler/docker image.

**Comment by jeff-cohere** [comment]
Having trouble reproducing the error again. The build system seems to have some "emergent features."

I don't know if we can reproduce this easily without PETSc. But if I can get it to show again, I'll try to build it without the PETSc environment set.

**Comment by jeff-cohere** [comment]
And don't worry about response times--I've got plenty of things to do! :-)

**Comment by jeff-cohere** [comment]
Okay, I managed to reproduce the ICE on my system again (the build system needs more work before it's reliable with dependencies).

The Ubuntu toolchain testing PPA doesn't include a version of GCC with this fix, which makes it a bit more complicated to try with our MPI setup. However, the same link we were looking at mentions that GFortran 8.4 doesn't have this issue. So another option for us would be to use an older version of GCC in a Docker image. This is not an exciting or welcome option, though--particularly if the PETSc Docker image is being used by others.

The approach I was taking was to exclude code coverage flags on the steady and transient F90 demos. That "works", except that the transient tests don't pass their benchmarks. I've run Valgrind on that executable and it gives me a clean bill of health. I'll keep poking.

**Comment by jeff-cohere** [comment]
For whatever reason, the `transient-mpfaof90-tpf` and `transient-mpfaof90-tpf-neumann` tests are failing when the codecov flags are omitted from the the `transient_mpfaof90` demo exe. It's unclear to me whether these are just sensitive tests that are affected by the different runtime environment, or whether something else is going on.

Here are the details (I made `regression_tests.py` dump log files for test suites that contain one or more failed tests):

```
<snip>
transient-mpfaof90-tpf... 
    cd /__w/TDycore/TDycore/demo/transient
    /__w/TDycore/TDycore/demo/transient/transient_mpfaof90 -malloc 0 -successful_exit_code 0 -ts_dt 100 -ts_type beuler -ts_max_time 1000 -ts_max_steps 10 -tdy_regression_test -tdy_regression_test_num_cells_per_process 2 -tdy_regression_test_filename transient-mpfaof90-tpf -ts_monitor -snes_monitor -ksp_monitor -ts_view -pc_type lu -tdy_mpfao_gmatrix_method MPFAO_GMATRIX_TPF
    # transient-mpfaof90-tpf : run time : 0.10 seconds
    diff transient-mpfaof90-tpf.regression.gold transient-mpfaof90-tpf.regression
    FAIL: Liquid Pressure:Max : 1.7872448996569385e-09 > 1e-12 [relative]
    FAIL: Liquid Pressure:Min : 2.6041672514028017e-10 > 1e-12 [relative]
    FAIL: Liquid Pressure:Mean : 3.522507231360993e-11 > 1e-12 [relative]
    FAIL: Liquid Pressure:0 : 9.163682765787368e-09 > 1e-12 [relative]
    FAIL: Liquid Pressure:13 : 1.2985290589314483e-07 > 1e-12 [relative]
transient-mpfaof90-tpf... failed.
----------------------------------------
transient-mpfaof90-tpf-neumann... 
    cd /__w/TDycore/TDycore/demo/transient
    /__w/TDycore/TDycore/demo/transient/transient_mpfaof90 -malloc 0 -successful_exit_code 0 -ts_dt 100 -ts_type beuler -ts_max_time 1000 -ts_max_steps 10 -tdy_regression_test -tdy_regression_test_num_cells_per_process 2 -tdy_regression_test_filename transient-mpfaof90-tpf-neumann -ts_monitor -snes_monitor -ksp_monitor -ts_view -pc_type lu -tdy_mpfao_gmatrix_method MPFAO_GMATRIX_TPF -tdy_mpfao_boundary_condition_type MPFAO_NEUMANN_BC
    # transient-mpfaof90-tpf-neumann : run time : 0.10 seconds
    diff transient-mpfaof90-tpf-neumann.regression.gold transient-mpfaof90-tpf-neumann.regression
    FAIL: Liquid Pressure:Max : 3.3433560808083567e-09 > 1e-12 [relative]
    FAIL: Liquid Pressure:Min : 7.01748114996301e-09 > 1e-12 [relative]
    FAIL: Liquid Pressure:Mean : 3.524692201775936e-11 > 1e-12 [relative]
    FAIL: Liquid Pressure:0 : 7.607709206182365e-09 > 1e-12 [relative]
    FAIL: Liquid Pressure:13 : 1.9411991976954824e-07 > 1e-12 [relative]
transient-mpfaof90-tpf-neumann... failed.
<snip>
```

**Comment by jeff-cohere** [comment]
Hey @ghammond86 ,

See the comment above this one for a demonstration of the "gold diff" issue you encountered in PFlotran with GCC/GFortran 10+. As you point out, changing the gold files will make these tests fail for folks using older compilers. This is one of the many problems associated with using these kinds of tests.

Do you have any thoughts on this? How is PFlotran dealing with these kinds of issues?

**Comment by ghammond86** [comment]
> Hey @ghammond86 ,
> 
> See the comment above this one for a demonstration of the "gold diff" issue you encountered in PFlotran with GCC/GFortran 10+. As you point out, changing the gold files will make these tests fail for folks using older compilers. This is one of the many problems associated with using these kinds of tests.
> 
> Do you have any thoughts on this? How is PFlotran dealing with these kinds of issues?

We are sticking with building on Ubuntu 18 for now with its GNU 7.X.  As long as code modifications are passing the automated builds (using GNU 7.X), I am not concerned, as long as all the failures are "F" or "M".  The updated version of PFLOTRAN regression testing returns a character for failed tests:

Running pflotran regression tests :
  Legend
    . - success
    F - failed regression test (results are outside error tolerances)
    M - failed regression test (results are FAR outside error tolerances)
    G - general error
    U - user error
    V - simulator failure (e.g. failure to converge)
    X - simulator crash
    T - time out error
    C - configuration file [.cfg] error
    I - missing information (e.g. missing files)
    B - pre-processing error (e.g. error in simulation setup scripts
    A - post-processing error (e.g. error in solution comparison)
    S - test skipped
    W - warning
    ? - unknown

**Comment by jeff-cohere** [comment]
Thanks, Glenn. @bishtgautam , do you have a preferred approach for TDycore in the light of all this nonsense?

**Comment by jedbrown** [comment]
What's the most recent ubuntu with a stable gcc that is not buggy (I guess that's 8.x)? I can spin up an image based on that.

**Comment by jeff-cohere** [comment]
Sounds like Ubuntu 18, maybe. I think that's a good call.

**Comment by jeff-cohere** [comment]
I'm closing this PR and continuing work in this branch to implement our own Docker PETSc image build (as we discussed yesterday).

**Pull request closed without merging**